File: src/lib/software/vlc_media_player.py
from re import compile
from PySide6.QtCore import Slot, QUrl
from src.lib.software import BaseSoftware, SoftwareCategory
from PySide6.QtNetwork import QNetworkReply, QNetworkRequest

class VlcMediaPlayer(BaseSoftware):
    def __init__(self):
        super().__init__()

        self.key = 'vlc-media-player'
        self.name = 'VLC Media Player'
        self.category = SoftwareCategory.Media
        self.download_name = 'vlc-win64.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = True
        self.icon = 'vlc-media-player.png'
        self.homepage = 'https://videolan.org'

        self._initial_url = QUrl('https://www.videolan.org/')

    @Slot(QNetworkReply)
    def on_manager_finished(self, reply: QNetworkReply):
        reply.deleteLater()
        error = reply.error()
        if error != QNetworkReply.NetworkError.NoError:
            self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            return

        html = reply.readAll().data().decode()

        if reply.url() == self._initial_url:
            download_page_pattern = compile(r'//get\.videolan\.org/vlc/\d+\.\d+\.\d+/win64/vlc-\d+\.\d+\.\d+-win64\.exe')
            match = download_page_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                req = QNetworkRequest(f'https:{match.group(0)}')
                self.manager.get(req)
        else:
            download_url_pattern = compile(r'URL=\'(.*)\'')
            match = download_url_pattern.search(html)
            if not match:
                self.url_resolve_error.emit(self.ResolveError.URLResolveError)
            else:
                self.url_resolved.emit(match.group(1))

    def resolve_download_url(self):
        self.manager.get(QNetworkRequest(self._initial_url))

File: packages/app/src/services/software/definitions/vlcMediaPlayer.ts
import { ok, err } from 'neverthrow';
import { BROWSER_USER_AGENT } from '~/constants';
import { SoftwareCategory, DownloadUrlResolveError } from 'shared';
import type { ISoftwareDefinition } from 'shared';

export default class implements ISoftwareDefinition {
	public key = 'vlc-media-player';
	public name = 'VLC Media Player';
	public category = [SoftwareCategory.Audio, SoftwareCategory.Media];
	public downloadName = 'vlc-win64.exe';
	public isArchive = false;
	public shouldCacheUrl = true;
	public requiresAdmin = true;
	public deprecated = false;
	public icon = 'vlc-media-player.png';
	public homepage = 'https://videolan.org';

	public async resolveDownloadUrl() {
		const downloadPageUrl = await this.getDownloadPageUrl();
		if (!downloadPageUrl) {
			return err(DownloadUrlResolveError.Generic);
		}

		const res = await fetch(downloadPageUrl, {
			headers: {
				'user-agent': BROWSER_USER_AGENT
			}
		});
		if (!res.ok) {
			return err(DownloadUrlResolveError.HTTPResponseError);
		}

		const html  = await res.text();
		const match = html.match(/URL=\'(.*)\'"/);
		if (!match) {
			return err(DownloadUrlResolveError.Generic);
		}

		return ok(match[1]);
	}

	private async getDownloadPageUrl() {
		const res = await fetch('https://www.videolan.org/');
		if (!res.ok) {
			return null;
		}

		const html  = await res.text();
		const match = html.match(/\/\/get\.videolan\.org\/vlc\/\d+\.\d+\.\d+\/win64\/vlc-\d+\.\d+\.\d+-win64\.exe/);
		if (!match) {
			return null;
		}

		return `https:${match[0]}`;
	}
}
